---
title: "Making R easier with Tidyverse<html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html>"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "https://rockefelleruniversity.github.io/RU_tidyverse/"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of contents true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  isSlides: "no"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

knitr::opts_knit$set(root.dir = "RU_tidyverse/inst/extdata/")
```

```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides != "yes"){
  cat("# Making R easier with Tidyverse

---
"    
  )
  
}

```

## Data wrangling with the Tidyverse

<b>There are some problems we face often when dealing with data.</b>

<ul>
  <li>Every dataset is different. Sometimes very different.</li>
  <li>There are many ways to do things. Everyone has their favorite syntax.</li>
</ul>

<b>The issue:</b>
Many fundamental data processing functions exist in <i>Base R</i> and beyond. Sometimes they can be inconsistent or unnecessarily complex. Especially when dealing with non-standard dataframes. The result is code that is confusing and doesn't flow i.e. nested functions

---
## What does it mean to be tidy?

Tidyverse is most importantly a philosophy for data analysis that more often then not makes wrangling data easier. The tidyverse community have built what they describe as an <i>opinionated</i> group of packages. These packages readily talk to one another. 

<ul>
  <li>More efficient code</li>
  <li>Easier to remember syntax</li>
  <li>Easier to read syntax</li>
</ul>

You can read their <a href="https://cran.r-project.org/web/packages/tidyverse/vignettes/manifesto.html">manifesto</a> to get a better understanding of the tidy ethos. 
  
---
## What does it <i>actually</i> mean to be tidy?

<ul>
  <li>A defined vision for coding style in R  
  <li>A defined vision for data formats in R
  <li>A defined vision for package design in R
  <li>Unified set of community pushing in a cohesive direction
  <li>Critical mass of people to influence the way the whole R community evolves
</ul>

---
## What are the main tidy tools?

<ul>
  <li>ggplot2 – making pretty graphs 
  <li>readr – reading data into R  
  <li>dplyr – manipulating data  
  <li>tibble - working with tibbles  
  <li>tidyr – miscellaneous tools for tidying data
  <li>purrr - iterating over data  
  <li>stringr – working with strings  
  <li>forcats - working with factors  
</ul>

Other tools have now been made by the Tidy community. This community also overlaps with Bioconductor. But the packages above are the linchpins that hold it together. 

---
## What we won't be doing today

We already have a course online for plotting, including ggplot.     

https://rockefelleruniversity.github.io/Plotting_In_R/

---
# Lets get tidy!
First step lets load in the data we are using today
```{r}

load(file='data/my_tidy.Rdata')
```

---
## Are all data frames equal?
```{r}
head(df1)
```

```{r}
head(df2)
```

---
## Are all data frames equal?
```{r}
head(df3a)
```

```{r}
head(df3b)
```

---
## What is a tidy dataset?

A tidy dataset is a data frame (or table) for which the following are true:

* Each variable has its own column
* Each observation has its own row
* Each value has its own cell

***Which of our dataframes is tidy?***
--
<p>&nbsp;</p>
Our first dataframe is tidy

---
## Why bother?

Consistent dataframe layouts help to ensure that all values are present and that relationships between data points are clear.

R is a vectorized programming language. R builds data frames from vectors, and R works best when its operation are vectorized.

Tidy data utilizes both these aspects of R.  
=> Precise and Fast


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Reading and Tibbles

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Reading and Tibbles

---
"    
  )
  
}

```


## Lets load in the tidyverse
```{r}
library(tidyverse)
```



---
## readr: Reading data into R

Lets start from the beginning and tidy some data. First step is to read in data. 

readr:

* read_csv(): comma separated (CSV) files  
* read_tsv(): tab separated files  
* read_delim(): general delimited files  
* read_fwf(): fixed width files  
* read_table(): tabular files where columns are separated by white-space  
* read_log(): web log files   

---
## Reading in with base

When you read data in with base, it goes into a regular dataframe. When you return the dataframe through typing it in the console it will just print the whole dataframe.

```{r}
untidy_counts_base <- read.csv("data/hemato_rnaseq_counts.csv")

untidy_counts_base
```

---
## Reading in with readr

When you use readr to read in your data, you instead get a special dataframe called a tibble. Tibbles have several properties that make them more user friendly i.e. When you return the tibble through typing it in the console it will print a preview, with some information about dimensions and data types. 
```{r}

read_csv("data/hemato_rnaseq_counts.csv")

```

---
## Reading in with readr
Tibbles carry and display extra information. While reading in it is easy to specify data type. 
```{r}
untidy_counts <- read_csv("data/hemato_rnaseq_counts.csv", col_types = cols(
    ENTREZ = col_character(),
    CD34_1 = col_integer(),
    ORTHO_1 = col_integer(),
    CD34_2 = col_integer(),
    ORTHO_2 = col_integer()
  ))
untidy_counts
```

## Subsetting to make tibbles
You can use the same methods as base to interact with tibbles to subset them.

```{r}
untidy_counts[1,]
```
```{r}
untidy_counts[,1]
```

---
## Subsetting to make tibbles
Tibbles are column-oriented, so if you don't specify row/column with a comma, it will assume you want a column. 
```{r}

untidy_counts[1]
```

---
## Subsetting to make vectors
The outputs thus far have produced more tibbles. Sometimes we just want a vector. Like lists, we can use double brackets. 
```{r}
untidy_counts[[1]]
```

---
## Subsetting to make vectors
Also, like lists we can use a dollar sign with the column name.
```{r}
untidy_counts$ENTREZ
```

---
## Tibbles: Converting to tibble
There are coercion functions similar to base to convert a regular dataframe to a tibble. 
```{r}

as_tibble(untidy_counts_base)
```

---
## Tibbles: Converting to tibble
Once it is a tibble it is straight forward to modify the datatype using the dplyr function mutate_at. 
```{r}

untidy_counts_base <- as_tibble(untidy_counts_base)
untidy_counts_base <- mutate_at(untidy_counts_base, vars(ENTREZ), as.character)
untidy_counts_base
```

---
## Tibbles: Converting from tibble

Some tools are not tibble friendly. Calling as.data.frame is sufficient to convert it back to a base data frame

```{r}

as.data.frame(untidy_counts_base) %>% head(n=12)
```

---
## Tibbles: Make your own - (pt1)
We will make our own tibble now from scratch, using some metadata. We are accessing some Bioconductor databases that contain annotation information for genes. 
First we grab the position of genes. Then we use the IDs to get gene symbols. We will cover how to do this in more detail later on [here](https://rockefelleruniversity.github.io/Bioconductor_Introduction/presentations/singlepage/GenomicFeatures_In_Bioconductor.html#Extracting_Genomic_Locations)

```{r}
# Lets load in some packages
library(org.Hs.eg.db)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)

hg19_genes <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)

keys <- hg19_genes$gene_id

symbols <- AnnotationDbi::select(org.Hs.eg.db, keys = keys, columns = c("SYMBOL"), keytype = "ENTREZID")

```


---
## Tibbles: Make your own
Now we will put all the metadata together into a tibble
```{r}

counts_metadata <- tibble(ID = symbols$ENTREZID, SYMBOL = symbols$SYMBOL, LENGTH = lengths(hg19_genes))

counts_metadata
```


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Tidying up your data

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Tidying up your data

---
"    
  )
  
}

```


## Tidying data up

What is wrong with the count dataframe from a tidy viewpoint?  

_Remember_. These are the rules:

* Each variable has its own column
* Each observation has its own row
* Each value has its own cell

```{r}
untidy_counts
```

---
## Tidying data up

What is wrong with the count dataframe from a tidy viewpoint?  

_Remember_, these are the rules:

* Each variable has its own column
* Each observation has its own row
* Each value has its own cell

__A single variable with multiple columns__

```{r}
untidy_counts
```

---
## Reshape data with gather and spread

When all the content is in the tibble, but it is in the wrong orientation, tidyr has some tools to move the data around quickly and easily:  

__gather and spread__

---
## tidyr::pivot_longer
pivot_longer allows you to collapse single variables that are spread over multiple columns. In this case new columns Sample and Counts are created. Sample is made from the column names hence *names_to*. While Counts is made from the values within the columns hence *values_to*. We are also specifying that we are tidying all columns apart from ENTREZ with the cols argument. 

**This recently replaced a function called gather**
```{r}

tidier_counts <- pivot_longer(untidy_counts, names_to = "Sample", values_to = "counts", cols = c(-ENTREZ))
tidier_counts

```

---
## tidyr::pivot_wider
pivot_wider allows you to go in the oopsite direction to pivot_longer. We can spread single variables over multiple columns.
```{r}
pivot_wider(tidier_counts, names_from = c(Sample), values_from = counts)
```

---
## What is next to tidy?

_Remember_, these are the rules:

* Each variable has its own column
* Each observation has its own row
* Each value has its own cell

```{r}
tidier_counts
```

---
## What is next to tidy?

_Remember_, these are the rules:

* Each variable has its own column
* Each observation has its own row
* Each value has its own cell

__Multiple variables in a single column__

```{r}
tidier_counts
```

---
## Splitting and combining varaibles

When there are several variables crammed into a single column, tidyr can be used to split single values into several:  

__separate__

---
## tidyr::separate
Separate allows you to break a strings in a variable by a separator. In this case the cell type and replicate number are broken by underscore.
```{r}
tidy_counts <- separate(tidier_counts, Sample, sep = "_", into=c("CellType", "Rep"), remove=TRUE)
tidy_counts
```

---
## tidyr::unite
Unite can go the other way. This can sometime be useful i.e. if you want a specific sample ID
```{r}
unite(tidy_counts, Sample, CellType, Rep, remove=FALSE)
```


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Piping with Magrittr

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Piping with Magrittr

---
"    
  )
  
}

```


## Simplifying code

Often when getting started code can get a little out of control and confusing. As addiotnal steps are adeed code can become nested. For example here we tidy our data in a single complicated expression 

```{r}
tidy_counts <- separate(pivot_longer(untidy_counts, names_to = "Sample", values_to = "counts", cols = c(-ENTREZ)), Sample, sep = "_", into = c("CellType","Rep"), remove=FALSE)
```

We simplified things by saving each step as an intermediate. This is not very efficient. 

```{r}
tidier_counts <- pivot_longer(untidy_counts, names_to = "Sample", values_to = "counts", cols = c(-ENTREZ))
tidy_counts <- separate(tidier_counts, Sample, sep = "_", into = c("CellType","Rep"), remove=FALSE)

```


---
## Piping to string functions together

Piping allows you to pass the result from one expression directly into another.  
magrittR package developed the %>% pipe which is integral to the tidy way of formatting code

The pattern is similar but now follows a specific logical flow:

![dplyr](./imgs/dplyr_pipe2.png)

## Piping to string functions together
This way there are no intermediates and the steps in your data analysis are logical and clear unlike nested code. 
```{r}

tidy_counts <- untidy_counts %>% 
  gather(key=Sample, value=counts, -ENTREZ) %>% 
  separate(Sample, sep = "_", into = c("CellType","Rep"), remove=FALSE)
tidy_counts
```


## Advanced Piping

A period ( . ) can be used as placeholders to represent the object being piped in. 
In this case we use the period to in a logical expression, which we then use to subset itself. 
```{r}

tidier_counts %>% .[.$counts > 0,]

```

%<>% is a modification of the pipe. This is a two way pipe that will carry the variable forward, and once expression is resolved, the result then gets assigned back to the orgonal variable.

````{r}
tidier_counts %<>% .[.$counts > 0,]
tidier_counts
```

Keep your eyes out for a base R pipe as it is under development at the moment *|>*

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Joining tibbles together

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Joining tibbles together

---
"    
  )
  
}

```


---
## Joining
Often you have two dataframes and you want to join them together. Data frames can be joined on a shared variable _a.k.a._ a key. We want this key to be unique i.e. ENTREZ ID. 

---
## dplyr::inner_join - Merging dataframes
```{r}
tidy_counts
```
---
## dplyr::inner_join - Merging dataframes
```{r}
counts_metadata
```
---
## dplyr::inner_join - Merging dataframes
```{r}
inner_join(tidy_counts, counts_metadata, by = c("ENTREZ" = "ID"))
```

---
## There are many ways to join things

Inner Join 

* Keeps all observations in x and y with matching keys

Outer Join

* A left join keeps all observations in x and those in y with matching keys.
* A right join keeps all observations in y and those in x with matching keys. 
* A full join keeps all observations in x and y

---

---
## dplyr::left_join 
My tidy_counts tibble is the on the left, so is the backbone for this join. This will only then show the metadata that has a matching key i.e. ENTREZID. 
```{r}
left_join(tidy_counts, counts_metadata, by = c("ENTREZ" = "ID"))
```

---
## dplyr::right_join
Now the counts_metadata tibble is the backbone for this join. Many of the keys (i.e. ENTREZID) in this tibble do not have matches in the tidy_counts_tibble. Anything without a match is filled in with NA. 
```{r}
tidy_counts_expressed <- right_join(tidy_counts, counts_metadata, by = c("ENTREZ" = "ID"))
tail(tidy_counts_expressed)
```

---
## Filtering joins

Filtering joins 

* A semi join only keeps all observations in x that are matched in y. y isn't returned. 
* A anti join only keeps all observations in x that are _not_ matched in y. y isn't returned. 

---
## dplyr::semi_join
Semi join only keeps observations in x that are matched in y. y is only used as a reference and is not in output
```{r}
semi_join(counts_metadata, tidy_counts, by = c("ID" = "ENTREZ") )
```
---
## dplyr::anti_join
Anti join only keeps observations in x that are not matched in y. y is only used as a reference and is not in output
```{r}

anti_join(counts_metadata, tidy_counts, by = c("ID" = "ENTREZ") )

```

---
## CHALLENGE: Time for an exercise!

Exercise reviewing dplyr and joining can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/exercises/EX7_RU_tidyverse_dplyr_join.html)

---
### Answers to exercise.

Answers can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/answers/EX7_ANSWERS_RU_tidyverse_dplyr_join.html)

R code for solutions can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/answers/EX7_ANSWERS_RU_tidyverse_dplyr_join.R)

---
# dplyr
This package contains a variety of tools to access and manipulate dataframes.  

They have a common rationale:

![dplyr](./imgs/dplyr.png)

---
## dplyr::select - (pt1)
**Select** allows you to make a vector from a specific variable
```{r}
# Select one variable (common_name)
select(df1, common_name)
```

---
## dplyr::select - (pt2)
**Select** allows you to make a dataframe from several variables
```{r}
# Select two variables (age_classbylength and common_name)
select(df1, age_classbylength, common_name)
```

---
## dplyr::select - (pt3)
**Select** allows you to make a dataframe excluding a variable
```{r}
# Select all but one variable (length_mm)
select(df1,-length_mm)
```

---
## dplyr::select - (pt4)
**Select** allows you to make a dataframe from a range of variables
```{r}
# Select all a range of contiguous varibles (common_name:length_mm)
select(df1, common_name:length_mm)
```

---
## dplyr::filter - (pt1)
__Filter__ allows you to access observations based on a criteria
```{r}
# Filter all observation where the variable common_name is Sockeye salmon
filter(df1, common_name == 'Sockeye salmon')
```

---
## dplyr::filter - (pt2)
__Filter__ allows you to access observations based on several criteria
```{r}
# Filter all observations where the variable common_name is either Sockeye salmon or Chinook Salmon
filter(df1, common_name %in% c('Sockeye salmon', 'Chinook salmon'))
```

---
## dplyr::filter - (pt3)
__Filter__ allows you to access observations based on specific strings
```{r}
# Filter all observations where the variable common_name ends with 'salmon'. To do this we use stringr function str_ends recognise strings that end with 'salmon'.
filter(df1, str_ends(common_name, 'salmon'))
```

---
## dplyr::filter - (pt4)
__Filter__ allows you to access observations based on operators
```{r}
# Filter all observations where the variable length_mm is greater than 200 or less than 120
filter(df1, length_mm > 200 | length_mm < 120)
```

---
## CHALLENGE: Time for an exercise!

Exercise on dplyr's select and filter can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/exercises/EX1_RU_tidyverse_dplyr_select_and_filter.html)

---
### Answers to exercise.

Answers can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/answers/EX1_ANSWERS_RU_tidyverse_dplyr_select_and_filter.html)

R code for solutions can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/answers/EX1_ANSWERS_RU_tidyverse_dplyr_select_and_filter.R)

---
## dplyr::arrange - (pt1)
_Arrange_ sorts the dataframe based on a specific variable
```{r}
# Arrange the data based on the variable length_mm
arrange(df1, length_mm)
```
---
## dplyr::arrange - (pt2)
_Arrange_ sorts the dataframe based on specific variables
```{r}
# Arrange the data first based on the variable common_name, 
# then secondly based on length_mm in a descending order.
arrange(df1, common_name, desc(length_mm))
```

---
## dplyr::mutate - (pt1)
_Mutate_ creates a new variable based on some form of computation
```{r}
# A new variable is created based on the calculation of the
# z-score of the variable IGF1_ng_ml using scale()
mutate(df1, scale(IGF1_ng_ml))
```
---
## dplyr::mutate - (pt2)
_Mutate_ creates a named variable based on some form of computation
```{r}
# A new variable is created called IGFngml_zscore, based on the 
# calculation of the z-score of the variable IGF1_ng_ml using scale()
mutate(df1, IGFngml_zscore = scale(IGF1_ng_ml))
```

---
## CHALLENGE: Time for an exercise!

Exercise on dplyr's arrange and mutate can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/exercises/EX2_RU_tidyverse_dplyr_arrange_and_mutate.html)

---
### Answers to exercise.

Answers can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/answers/EX2_ANSWERS_RU_tidyverse_dplyr_arrange_and_mutate.html)

R code for solutions can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/answers/EX2_ANSWERS_RU_tidyverse_dplyr_arrange_and_mutate.R)

---
## dplyr::summarize - (pt1)
_Summarize_ applies aggregating or summary function to a group i.e. counting
```{r}
# First we define the common_name as a group. 
df1_byname <- group_by(df1, common_name)

# Summarise is used to count over the grouped common_names
summarise(df1_byname, count = n())
```
---
## dplyr::summarize - (pt2)
_Summarize_ applies aggregating or summary function to a group i.e. means
```{r}
# Summarise is used to calculate mean IGF1_ng_ml over the
# grouped common_names
summarise(df1_byname, IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm = T))
```

---
## dplyr::group_by - (pt1)
_Grouping_ can also help filter within groups
```{r}
# Filter observations with the 2 smallest length_mm for each grouped common_names
filter(df1_byname, rank(length_mm) <= 2)
```
---
## dplyr::group_by - (pt2)
_Grouping_ can also help filter within groups
```{r}
# Filter observations with at least 5 for each grouped common_names
filter(df1_byname, n() > 5)
```
---
## dplyr::group_by - (pt3)
_Grouping_ creates a new variable based on some form of computation within the group
```{r}
# A new variable is created using z-score within the grouped common_names
mutate(df1_byname, IGFngml_zscore = scale(IGF1_ng_ml))
```

---
## CHALLENGE: Time for an exercise!

Exercise on dplyr's grouping and summarize can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/exercises/EX3_RU_tidyverse_dplyr_group_and_summarize.html)

---
### Answers to exercise.

Answers can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/answers/EX3_ANSWERS_RU_tidyverse_dplyr_group_and_summarize.html)

R code for solutions can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/answers/EX3_ANSWERS_RU_tidyverse_dplyr_group_and_summarize.R)

---
# Piping to string functions together

Piping was allows you to pass the result from one expression directly into another.  
magrittR package developed the %>% pipe which is integral to the tidy way of formatting code

The pattern is similar but now follows a specific logical flow:

![dplyr](./imgs/dplyr_pipe.png)
---
## Piping vs. not piping
```{r}
# Without pipe
df1_byname <- group_by(df1, common_name)
summarise(df1_byname, IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm=T))

# With pipe
df1 %>% 
  group_by(common_name) %>% 
  summarise(IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm=T))
```

---
## Linking functions with pipes - (pt1)
```{r}
# (1) Group by common_name
# (2) Filter to all those that have length bigger then 200
# (3) Summarise is used to calculate mean IGF1_ng_ml over the grouped 
# common_names for these larger fish

df1 %>%
  group_by(common_name) %>% 
  filter(length_mm > 200) %>% 
  summarise(IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm = T))
```
---
## Linking functions with pipes - (pt2)
```{r}
# (1) Create new variable that is discrete label depending on size of the fish
# (2) Group by common_name and size
# (3) Summarise is used to calculate mean IGF1_ng_ml over the grouped 
# common_names and sizes

df1 %>%
  mutate(size = if_else(length_mm > 200, 'big_fish', 'small_fish')) %>% 
  group_by(common_name, size) %>% 
  summarise(IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm = T))
```
---
## Linking functions with pipes - (pt3)
```{r}
# (1) Create new variable that is discrete label depending on size of the fish
# (2) Group by common_name and size
# (3) Summarise is used to calculate mean IGF1_ng_ml over the grouped 
# common_names and sizes
# (4) Filter out Coho and Sockeye salmon

df1 %>%
  mutate(size = if_else(length_mm > 200, 'big_fish', 'small_fish')) %>% 
  group_by(common_name, size) %>% 
  summarise(IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm = T)) %>%
  filter(common_name != 'Coho salmon')  %>%
  filter(common_name !='Sockeye salmon')
```

---
## You can pipe straight to plots - (pt1)
```{r, warning=F, message=F}

p <- df1 %>% mutate(size=if_else(length_mm>200, 'big_fish', 'small_fish')) %>% 
  group_by(common_name, size) %>% 
  summarize(IGF1_ng_ml_ave=mean(IGF1_ng_ml, na.rm=T)) %>% 
  filter(common_name != 'Coho salmon')  %>% 
  filter(common_name != 'Sockeye salmon') %>% 
  ggplot(aes(x = common_name, y = IGF1_ng_ml_ave, group = size, fill = size)) +
  geom_bar(stat = "identity", position = "dodge") + 
  theme(axis.text.x = element_text(angle = 90)) +
  scale_fill_brewer(palette = "Paired")

```
---
## You can pipe straight to plots - (pt2)
```{r, warning=F, message=F}
p
```



---
## CHALLENGE: Time for an exercise!

Exercise on piping can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/exercises/EX4_RU_tidyverse_pipes.html)

---
### Answers to exercise.

Answers can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/answers/EX4_ANSWERS_RU_tidyverse_pipes.html)

R code for solutions can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/answers/EX4_ANSWERS_RU_tidyverse_pipes.R)



---
## CHALLENGE: Time for an exercise!

Exercise on readr can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/exercises/EX5_RU_tidyverse_readr.html)






---
# Readr again: Writing out 
You've made a lovely new tibble file. Now you need to save it somewhere. 
```{r}
#Theres a wide range of writing  options. Can specify the delmiter directly or use a specific function
write_delim(tidy_counts_expressed_norm, '../expressed_genes_output.csv', delim =',')

write_csv(tidy_counts_expressed_norm, '../expressed_genes_output.csv')
```

A key difference compared to base is that it does not write out row names. Tibbles generally don't have rownames.

---
## tidyverse so far
__At this point we have covered or touched on the most essential facets of tidy__  
* ~~ggplot2 – making pretty graphs~~ 
* ~~readr – reading data into R~~  
* ~~dplyr – manipulating data~~  
* ~~tibble - working with tibbles~~  
* ~~tidyr – miscellaneous tools for tidying data~~
* purrr - iterating over data  
* stringr – working with strings  
* forcats - working with factors  
---
# stringr - working with strings

If the data you are working with involves characters from data entry often there will be errors i.e. clinical study metadata or a hand-typed list of genes of interest. Tidying data also means fixing these problems. stringr helps make this easy. 

* Access and manipulate characters
* Deal with whitespace
* Pattern Recognition

Though stringr is pretty comprehensive and covers most of what you will need, there is a sister package called stringi with even more functionality. 

---
## stringr::strsub - string basics
Many overlapping functions with base for combining, subsetting, converting and finding strings

```{r}

brc <- c("Tom", "Ji-Dung", "Matt")

# Extract substrings from a range. Here the 1st to 3rd character
brc %>% str_sub(1, 3)

# Extract substrings from a range. Here the 2nd to 2nd to last character
brc %>% str_sub(2, -2)


# Assign values back to substrings. Here the 2nd to 2nd to last character is replaced with X.
str_sub(brc, 2, -2) <- 'X'
brc
```

---
## stringr::str_trim - stripping whitespace

```{r}
brc2 <- c("Tom  ", "  Ji  -Dung", "Matt   ")

# Trim whitespace from strings
brc2 <- str_trim(brc2)
brc2 

# Can add whitespace to strings to get consistent length. Here all are 10 characters
str_pad(brc2, width=10, side='left')
```

---
## stringr::str\_to\_* - Capitalization - (pt1)
```{r}
# Lets reuse our counts tibble. pull from dplyr can be used to grab a tibble 
# column and make it into a vector
tidy_counts_expressed_norm %>% 
  pull(SYMBOL) %>%
  head()

# Here we pull our gene symbols from our tibble into a vector, and then convert 
# them into title style capitalization
tidy_counts_expressed_norm %>% 
  pull(SYMBOL) %>% 
  str_to_title() %>% 
  head()
```
---
## stringr::str\_to\_* - Capitalization - (pt2)
```{r}
# String manipulation functions can be used on tibbles using mutate. Here we convert
# gene symbols to title style capitalization
tidy_counts_expressed_norm %>% 
  mutate(SYMBOL = str_to_title(SYMBOL))
```
---
## stringr::str\_to\_* - Capitalization - (pt3)
```{r}
# Here we convert chromosome annotation to capitals
tidy_counts_expressed_norm %>% 
  mutate(CHR = str_to_upper(CHR))
```

---
## stringr::detect - which have pattern?
```{r}
# Find patterns in different ways
# Detect gives a T/F whether the pattern 'salmon' is present in vector
df1 %>% 
  pull(common_name) %>% 
  str_detect('salmon')
```
---
## stringr::subset - give me only patterns
```{r}
# Subset returns the match if the pattern 'salmon' is present in vector
df1 %>% 
  dplyr::pull(common_name) %>% 
  str_subset('salmon') 
```
---
## stringr::str_ends - ends with pattern?
```{r}
# Ends is similar to detect as it gives gives a T/F whether the pattern 'salmon' 
# is present in vector, but the pattern has to be at the end. 
df1 %>% 
  dplyr::pull(common_name) %>% 
  str_ends('salmon') 

df1 %>% 
  filter(str_ends(common_name,'salmon'))
```
---
## stringr::str_count - how many patterns?
```{r}
#Count gives you the total number of times your pattern appears in each chracter in the vector
df1 %>% 
  dplyr::pull(common_name) %>% 
  str_count('salmon')

df1 %>% 
  dplyr::pull(common_name) %>% 
  str_count('o')
```
---
## stringr::str_replace_all - replace a string
```{r}
#Replace
df1 %>% 
  dplyr::pull(common_name) %>%
  str_replace_all('Steelhead','Steelhead trout' )
```
---
## stringr::str_replace_all - replace a string
```{r}
df1 %>% 
  mutate(common_name = str_replace_all(common_name,'Steelhead','Steelhead trout' ))
```

---
## CHALLENGE: Time for an exercise!

Exercise reviewing purrr can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/exercises/EX9_RU_tidyverse_stringr.html)

---
### Answers to exercise.

Answers can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/answers/EX9_ANSWERS_RU_tidyverse_stringr.html)

R code for solutions can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/answers/EX9_ANSWERS_RU_tidyverse_stringr.R)



---
## CHALLENGE: Time for an exercise!

Exercise reviewing purrr can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/exercises/EX10_RU_tidyverse_forcats.html)

---
### Answers to exercise.

Answers can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/answers/EX10_ANSWERS_RU_tidyverse_forcats.html)

R code for solutions can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/answers/EX10_ANSWERS_RU_tidyverse_forcats.R)

---
# Summary 
We have covered the core packages in the tidyverse  

* ~~ggplot2 – making pretty graphs~~ 
* ~~readr – reading data into R~~  
* ~~dplyr – manipulating data~~  
* ~~tibble - working with tibbles~~  
* ~~tidyr – miscellaneous tools for tidying data~~
* ~~purrr - iterating over data~~
* ~~stringr – working with strings~~  
* ~~forcats - working with factors~~  

---
## Tidy beyond this workshop

Hadley Wickham (Chief Scientist at RStudio) is the driving force behind the tidyverse.  

Hadley wrote a paper about why he thinks tidy data is best: www.jstatsoft.org/v59/i10/paper.  

There is a lot of support for all things tidy at: https://www.tidyverse.org/  
(This includes really great cheat sheets for each tool)

---
## Tidy packages to check out:

plyranges: dplyr equivalent for working with ranges i.e. Granges

profileplyr: graphical and heatmap outputs for genomics

readxl: This package is very useful when you want to import Excel sheets in R

googledrive: Interact with your googledrive through R

lubridate and hms: Allow managin of calendar and time formats

magrittr: piping tools

broom: helps tidy up standard base function i.e. lm or t.test

tidymodels: A collection of tools for preparing for and validating model functions

---
## Other Good Resources

ggplot  
https://rockefelleruniversity.github.io/Plotting_In_R/

R for Data Science text book  
https://r4ds.had.co.nz/

Data Science with R text book   
http://garrettgman.github.io/

Biounformatics Shared Training - Intermediate R course   
https://bioinformatics-core-shared-training.github.io/r-intermediate/





